lib/kubeclient/informer.rb

```diff
@@ -21,22 +22,34 @@ def watch(&block)
 
     # not implicit so users know they have to `stop`
     def start_worker
+      @stopped = false
       @worker = Thread.new do
         loop do
-          fill_cache
-          watch_to_update_cache
-        rescue StandardError => e
-          # need to keep retrying since we work in the background
-          @logger&.error("ignoring error during background work #{e}")
-        ensure
-          sleep(1) # do not overwhelm the api-server if we are somehow broken
+          begin
+            fill_cache
+            watch_to_update_cache
+          rescue StandardError => e
+            # need to keep retrying since we work in the background
+            @logger&.error("ignoring error during background work #{e}")
+          ensure
+            sleep(1) # do not overwhelm the api-server if we are somehow broken
+          end
+          break if @stopped
         end
       end
-      sleep(0.01) until @cache
+      sleep(0.01) until @cache || @stopped
     end
 
     def stop_worker
-      @worker&.kill # TODO: be nicer ?
+      @stopped = true
+      [@waiter, @worker].compact.each do |thread|
+        begin
+          thread.run # cancel sleep so either the loop sleep or the timeout sleep are interrupted
+        rescue ThreadError
+          # thread was already dead
+        end
+        thread.join
```

Comment:
Similar question about `@waiter` possibly blocking for many minutes.  
Here it'd block the app calling `stop_worker` right?

- I'm thinking there is a subordinate relationship between worker thread -> single `watch_to_update_cache` run -> waiter thread. :thinking:  
  What if waiter threads did not refer to current instance variable `@watcher.finish` but closed over a lexically scoped reference to _their_ watcher?  Would then it be safe to leak unfinished waiter threads without .join-ing them?  Would it be a good idea?

- Waiter threads don't do much — are they safe to .kill?  Or is there another way to interrupt a sleep() early?

Reply:
> Or is there another way to interrupt a sleep() early?

You can use `Concurrent::Event` with a wait timeout to implement an interruptible sleep.

Reply:
before we killed them, but that made the tests brittle, so I like the joining since that makes anything that goes wrong more obvious (and re-raises exceptions too)

Reply:
From my testing `@watcher.each` can get blocked on the response stream. Even closing the `http_client` will not unblock the response body stream.

In that case `join` on the waiter thread that runs the `@watcher.each` internally will never return